governance/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+                    &execute_payload.target,
+                    execute_payload.eta,
+                    ProposalEventData {
+                        call_data: &execute_payload.call_data,
+                        value: &execute_payload.native_value,
+                    },
+                );
+            }
+        }
+    }
+
+    fn schedule_time_lock(
+        &self,
+        hash: &ManagedByteArray<KECCAK256_RESULT_LEN>,
+        mut eta: u64,
+    ) -> u64 {
+        require!(!hash.is_empty(), "Invalid time lock hash");
+
+        let time_lock_eta_mapper = self.time_lock_eta(hash);
+
+        require!(
+            time_lock_eta_mapper.is_empty(),
+            "Time lock already scheduled"
+        );
+
+        let minimum_eta =
+            self.blockchain().get_block_timestamp() + self.minimum_time_lock_delay().get();
+
+        if eta < minimum_eta {
+            eta = minimum_eta;
+        }
+
+        time_lock_eta_mapper.set(eta);
+
+        eta
+    }
+
+    fn cancel_time_lock(&self, hash: &ManagedByteArray<KECCAK256_RESULT_LEN>) {
+        require!(!hash.is_empty(), "Invalid time lock hash");
+
+        self.time_lock_eta(hash).clear();
+    }
+
+    fn finalize_time_lock(&self, hash: &ManagedByteArray<KECCAK256_RESULT_LEN>) -> u64 {
+        let eta = self.time_lock_eta(hash).take();
+
+        require!(!hash.is_empty() && eta != 0, "Invalid time lock hash");
+        require!(
+            self.blockchain().get_block_timestamp() >= eta,
+            "Time lock not ready"
+        );
+
+        eta
+    }
+
+    fn get_proposal_hash(
+        &self,
+        target: &ManagedAddress,
+        call_data: &ManagedBuffer,
+        native_value: &BigUint,
+    ) -> ManagedByteArray<KECCAK256_RESULT_LEN> {
+        let mut encoded = ManagedBuffer::new();
+
+        encoded.append(target.as_managed_buffer());
+        encoded.append(call_data);
+        encoded.append(&native_value.to_bytes_be_buffer());
+
+        self.crypto().keccak256(encoded)
+    }
+
+    fn only_governance(&self, source_chain: &ManagedBuffer, source_address: &ManagedBuffer) {
+        require!(
+            self.crypto().keccak256(source_chain) == self.governance_chain_hash().get()
+                && self.crypto().keccak256(source_address) == self.governance_address_hash().get(),
+            "Not governance"
+        );
+    }
+
+    #[callback]
+    fn execute_proposal_callback(
+        &self,
+        hash: &ManagedByteArray<KECCAK256_RESULT_LEN>,
+        eta: u64,
+        #[call_result] call_result: ManagedAsyncCallResult<MultiValueEncoded<ManagedBuffer>>,
+    ) {
+        match call_result {
+            ManagedAsyncCallResult::Ok(results) => {
+                self.execute_proposal_success_event(hash, results);
+            }
+            ManagedAsyncCallResult::Err(err) => {
+                // Let call be retried in case of failure, mostly because async call
+                // can fail with out of gas since it can be triggered by anyone
```

Comment:
Just so that this is clear, the async call can actually be any call to any contract, so we don't know the gas of that call before hand.
Here I have switched to using Async Promises in order to reserve some gas for this callback to make sure that it will not fail under any circumstances.